content/en/docs/Core concepts/baker_components.md

```diff
@@ -0,0 +1,135 @@
+---
+title: "baker.Components"
+date: 2020-10-29
+weight: 3
+description: >
+  baker.Components is the main object used to create a Baker topology
+---
+
+`baker.Components` is a struct that is used to configure the Baker topology, before running it.
+
+It contains the available components and some functions used by Baker to deal with records.
+
+```go
+type Components struct {
+	Inputs         []InputDesc
+    Filters        []FilterDesc
+	Outputs        []OutputDesc
+    Uploads        []UploadDesc
+    
+	Metrics        []MetricsDesc
+	User           []UserDesc
+    
+    ShardingFuncs  map[FieldIndex]ShardingFunc
+	Validate       ValidationFunc
+	CreateRecord   func() Record
+	FieldByName    func(string) (FieldIndex, bool)
+	FieldName      func(FieldIndex) string
+}
+```
+
+## Inputs, Filters, Outputs and Uploads
+
+These fields contain the list of components that are available to the topology.
+
+The [TOML configuration file](/docs/core-concepts/toml/) must specify components that are
+present in these lists.
+
+Both components already available to Baker or custom components can be set here.
+
+The following is an example of `baker.Components` configuration where:
+
+* **inputs** and **uploads** are those already included into Baker
+* only a custom **filter** is set
+* a custom **output** is added to the outputs included into Baker
+
+```go
+import (
+	"github.com/AdRoll/baker"
+	"github.com/AdRoll/baker/input"
+	"github.com/AdRoll/baker/output"
+	"github.com/AdRoll/baker/upload"
+)
+
+comp := baker.Components{
+    Inputs:        input.All,
+    Filters:       []baker.FilterDesc{MyCustomFilterDesc},
+	Outputs:       append(output.All, MyCustomOutputDesc...),
+	Uploads:       upload.All,
+}
+```
+
+## Metrics
+
+The list of available metrics bakends.
+
+As for the components explained in the previous paragraph, metrics can be imported from what
+is included into Baker or can be a custom implementation of the `baker.MetricsClient` interface
+(or a mix of the two options).
+
+For details about metrics, [see the dedicated page](/docs/core-concepts/metrics).
+
+## User
+
+This field contains the list of user-defined configurations, that are configurations not strictly
+used by the Baker topology but that a user can read from the Baker TOML configuration file and use
+anywhere in the code.
+
+To learn more about this topic, read the
+[dedicated section](/docs/core-concepts/toml/#user-defined-configurations) in the Pipeline
+configuration page.
+
+## ShardingFuncs
+
+This field is a dictionary of functions used to calculate how to shard records among the
+available output goroutines (see [Tuning concurrency](/docs/core-concepts/concurrency) for
+details about output concurrency).
```

Comment:
The sentence is not super clear IMHO, what do you think of something along these lines?
```suggestion
This field holds a dictionary associating field indices to hash functions. When sharding
is enabled, these hash functions are used to determine which shard a record is sent to.
```

Also, I'm not sure we should then go into such details about sharding here since we're in the `baker.Components` section, IMHO the rest of the ShardingFunc section should reside in the sharding guide and/or sharding core concept pages, we should keep it very close to what  the baker.Components field does.